### PR TITLE
fix: generate MAD_MULTI_NODE_RUNNER for Docker local deployment

### PR DESCRIPTION
**Pull request by coketaste**
## Summary

  - Docker local mode had no mechanism to set `MAD_MULTI_NODE_RUNNER`, unlike SLURM and K8s which generate it in their deployment layers. This caused distributed training scripts (e.g.
  Megatron-LM `train_7b.sh`) to fail with `Error: MULTI_NODE_RUNNER is not defined` on local runs.
  - Adds `_generate_local_launcher_command()` to `ContainerRunner` and calls it in `run_container()` after GPU resolution, generating the appropriate single-node launcher command
  (torchrun, deepspeed, etc.) and injecting it as the `MAD_MULTI_NODE_RUNNER` env var.
  - No-op for models that don't use `MAD_MULTI_NODE_RUNNER` — the env var is simply unused.

  ## Details

  SLURM sets `MAD_MULTI_NODE_RUNNER` in `job.sh.j2` and K8s sets it in `kubernetes_launcher_mixin.py`. Docker local had no equivalent, so any model script referencing
  `$MAD_MULTI_NODE_RUNNER` would fail.

  The fix reuses the already-resolved `launcher` variable in `run_container()` and maps it to the correct single-node command:
  - `torchrun`/`megatron`/`torchtitan` → `torchrun --standalone --nproc_per_node=N`
  - `deepspeed` → `deepspeed --num_gpus=N`
  - `vllm`/`sglang`/`primus` → empty (these manage their own process spawning)
  - Defaults to `torchrun` when `launcher` is a deployment-level value like `"docker"` or `"native"`

  Only sets the env var if the user hasn't already provided it via `docker_env_vars`.

  ## Test plan

  - [ ] Run a model that uses `$MAD_MULTI_NODE_RUNNER` (e.g. Megatron-LM train_7b) locally — verify it no longer fails with `MULTI_NODE_RUNNER is not defined`
  - [ ] Run a model that doesn't use `$MAD_MULTI_NODE_RUNNER` (e.g. HuggingFace script) — verify no regression
  - [ ] Verify SLURM/K8s paths are unaffected (change is scoped to `ContainerRunner.run_container()`)

### CODE REVIEW

File: src/madengine/execution/container_runner.py [hunk 1]
```diff
@@ -1089,7 +1114,29 @@ def run_container(
...
+        # For models that hardcode their own launcher (e.g. HuggingFace scripts
+        # calling torchrun directly), this env var is simply unused.
+        if "MAD_MULTI_NODE_RUNNER" not in self.context.ctx["docker_env_vars"]:
+            dist_launcher = launcher if launcher in (
+                "torchrun", "megatron", "megatron-lm", "torchtitan",
+                "deepspeed", "vllm", "sglang", "sglang-disagg", "primus",
+            ) else "torchrun"
+            runtime_ngpus = int(self.context.ctx["docker_env_vars"].get(
+                "MAD_RUNTIME_NGPUS", str(resolved_gpu_count)))
+            launcher_cmd = self._generate_local_launcher_command(dist_launcher, runtime_ngpus)
```
[the remark this answers is not in the document]
Comment (coketaste) [reply src/madengine/execution/container_runner.py:1140]: Resolved

File: src/madengine/execution/container_runner.py [hunk 2]
```diff
@@ -657,6 +657,31 @@ def get_cpu_arg(self) -> str:
         cpus = self.context.ctx["docker_cpus"].replace(" ", "")
         return f"--cpuset-cpus {cpus} "
 
+    def _generate_local_launcher_command(self, launcher_type: str, nproc_per_node: int) -> str:
+        """Generate distributed process launcher command for Docker local deployment.
+
+        Docker local is always single-node. This parallels
+        SlurmDeployer._generate_launcher_command() and
```
[the remark this answers is not in the document]
Comment (coketaste) [reply src/madengine/execution/container_runner.py:664]: the docstring has a wrong class name.

  At container_runner.py:664, the docstring says SlurmDeployer._generate_launcher_command(), but the actual class in
  deployment/slurm.py:31 is SlurmDeployment. The KubernetesLauncherMixin._generate_torchrun_command() reference on line 665 is
  correct (kubernetes_launcher_mixin.py:11,26).

  It's a one-word docstring typo — purely cosmetic, no runtime impact. Worth a quick fix for accuracy, but safe to skip if you're
  batching low-priority comments. Want me to apply the rename?

File: src/madengine/execution/container_runner.py [hunk 3]
```diff
@@ -1089,7 +1114,29 @@ def run_container(
...
+        # SLURM and K8s generate this in their deployment layers (slurm.py,
+        # kubernetes_launcher_mixin.py). Docker local has no such layer, so
+        # we generate it here after GPU resolution provides MAD_RUNTIME_NGPUS.
+        # Reuses the `launcher` variable already resolved at lines 327-372.
+        # Defaults to torchrun when launcher is a deployment-level value
+        # ("docker", "native") rather than a distributed launcher.
+        # For models that hardcode their own launcher (e.g. HuggingFace scripts
+        # calling torchrun directly), this env var is simply unused.
+        if "MAD_MULTI_NODE_RUNNER" not in self.context.ctx["docker_env_vars"]:
+            dist_launcher = launcher if launcher in (
```
[the remark this answers is not in the document]
Comment (coketaste) [reply src/madengine/execution/container_runner.py:1134]: Added (src/madengine/deployment/common.py):
  - _LAUNCHER_ALIASES dict — single source of truth for alternate spellings (currently just sglang_disagg → sglang-disagg).
  - canonicalize_distributed_launcher() — resolves aliases; returns unknown/empty values unchanged so callers retain validation control.

  Changed dispatch sites to call the helper instead of branching on alternate spellings:
  - slurm.py:350 — was == "sglang-disagg" or == "sglang_disagg".
  - k8s_template_context.py:342 — same pattern.
  - container_runner.py:1134 (local) — canonicalize before tuple membership check; also now prints a warning when an unrecognized launcher silently falls back to torchrun (the deeper bug
  Copilot's comment hinted at).

File: src/madengine/execution/container_runner.py [hunk 4]
```diff
@@ -1089,7 +1114,29 @@ def run_container(
...
+                "torchrun", "megatron", "megatron-lm", "torchtitan",
+                "deepspeed", "vllm", "sglang", "sglang-disagg", "primus",
+            ) else "torchrun"
+            runtime_ngpus = int(self.context.ctx["docker_env_vars"].get(
+                "MAD_RUNTIME_NGPUS", str(resolved_gpu_count)))
+            launcher_cmd = self._generate_local_launcher_command(dist_launcher, runtime_ngpus)
+            if launcher_cmd:
+                self.context.ctx["docker_env_vars"]["MAD_MULTI_NODE_RUNNER"] = launcher_cmd
```
[the remark this answers is not in the document]
Comment (coketaste) [reply src/madengine/execution/container_runner.py:1142]: Addressed the first ask: added TestGenerateLocalLauncherCommand in tests/unit/test_container_runner.py covering all launcher branches (torchrun-family, deepspeed, self-managed launchers, unknown → torchrun fallback), plus TestCanonicalizeDistributedLauncher in tests/unit/test_deployment.py for the new alias helper.
Skipping the second ask (test that user-supplied MAD_MULTI_NODE_RUNNER is preserved). The guard is a single if "MAD_MULTI_NODE_RUNNER" not in ... inside the ~200-line run_models_from_manifest method — testing it would require either heavy mocking of Docker/GPU resolution or a refactor extracting the env-var generation into its own method. Given the simplicity of the guard and the low regression risk, that's scope creep beyond this PR. Happy to revisit in a follow-up if you want the env-var generation extracted into a testable helper.

File: src/madengine/execution/container_runner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (coketaste) [reply src/madengine/execution/container_runner.py:664]: Fixed at container_runner.py:665 — SlurmDeployer → SlurmDeployment to match the actual class name in src/madengine/deployment/slurm.py.

File: src/madengine/execution/container_runner.py [hunk 5]
```diff
@@ -1089,7 +1114,35 @@ def run_container(
...
+                launcher = os.environ.get("MAD_LAUNCHER", "")
+            dist_launcher = launcher if launcher in (
+                "torchrun", "megatron", "megatron-lm", "torchtitan",
+                "deepspeed", "vllm", "sglang", "sglang-disagg", "primus",
+            ) else "torchrun"
+            runtime_ngpus = int(self.context.ctx["docker_env_vars"].get(
+                "MAD_RUNTIME_NGPUS", str(resolved_gpu_count)))
+            launcher_cmd = self._generate_local_launcher_command(dist_launcher, runtime_ngpus)
+            if launcher_cmd:
+                self.context.ctx["docker_env_vars"]["MAD_MULTI_NODE_RUNNER"] = launcher_cmd
```
[the remark this answers is not in the document]
Comment (coketaste) [reply src/madengine/execution/container_runner.py:1142]: Fixed at container_runner.py:1150 — MAD_MULTI_NODE_RUNNER is now always set (possibly to empty string) once we enter this branch, so downstream scripts under set -u won't fail when
  launcher is vllm/sglang/primus. The log message still only prints for non-empty launcher commands.

File: src/madengine/execution/container_runner.py [hunk 6]
```diff
@@ -1089,7 +1114,35 @@ def run_container(
...
+        # calling torchrun directly), this env var is simply unused.
+        if "MAD_MULTI_NODE_RUNNER" not in self.context.ctx["docker_env_vars"]:
+            launcher = ""
+            if self.additional_context:
+                launcher = self.additional_context.get("distributed", {}).get("launcher", "")
+            if not launcher and model_info.get("distributed"):
+                launcher = model_info["distributed"].get("launcher", "")
+            if not launcher:
+                launcher = os.environ.get("MAD_LAUNCHER", "")
+            dist_launcher = launcher if launcher in (
```
[the remark this answers is not in the document]
Comment (coketaste) [reply src/madengine/execution/container_runner.py:1134]: - container_runner.py: extracted the inline MAD_MULTI_NODE_RUNNER resolution block into a new helper _resolve_local_multi_node_runner_env(model_info, resolved_gpu_count). run_container
  now calls it in one line. Behavior identical.
  - test_container_runner.py: added TestResolveLocalMultiNodeRunnerEnv with 13 parametrized cases covering Copilot's three asks:
    a. Sets expected command for torchrun/megatron/megatron-lm/torchtitan/deepspeed
    b. Does NOT override user-provided docker_env_vars["MAD_MULTI_NODE_RUNNER"]
    c. vLLM/SGLang/SGLang-disagg/Primus set the var to "" (defined-but-empty, the intended behavior after the prior set -u fix)
    - Plus model_info fallback, unknown-launcher → torchrun, and MAD_RUNTIME_NGPUS precedence.
